bundle/artifacts/build.go

```diff
@@ -27,61 +29,146 @@ func (m *build) Name() string {
+				log.Debugf(ctx, "Patching ind=%d artifactName=%s Source=%s patchedWheel=%s", ind, artifactName, artifactFile.Source, patchedWheel)
+				diags = append(diags, extraDiags...)
+				if patchedWheel != "" {
+					a.Files[ind].Patched = patchedWheel
+				}
+				if extraDiags.HasError() {
+					break
+				}
 			}
-			log.Infof(ctx, "%s: Build succeeded, found %s", artifactName, wheels)
-		} else {
-			log.Infof(ctx, "%s: Build succeeded", artifactName)
-		}
-
-		// We need to expand glob reference after build mutator is applied because
-		// if we do it before, any files that are generated by build command will
-		// not be included into artifact.Files and thus will not be uploaded.
-		diags = diags.Extend(bundle.Apply(ctx, b, expandGlobs{name: artifactName}))
-		if diags.HasError() {
-			break
 		}
 	}
 
 	return diags
 }
+
+func doBuild(ctx context.Context, artifactName string, a *config.Artifact) error {
+	cmdio.LogString(ctx, fmt.Sprintf("Building %s...", artifactName))
+
+	var executor *exec.Executor
+	var err error
+	if a.Executable != "" {
+		executor, err = exec.NewCommandExecutorWithExecutable(a.Path, a.Executable)
+	} else {
+		executor, err = exec.NewCommandExecutor(a.Path)
+		a.Executable = executor.ShellType()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	out, err := executor.Exec(ctx, a.BuildCommand)
+	if err != nil {
+		return fmt.Errorf("build failed %s, error: %v, output: %s", artifactName, err, out)
+	}
+
+	return nil
+}
+
+func makePatchedWheel(ctx context.Context, cacheDir, artifactName, wheel string) (string, diag.Diagnostics) {
+	msg := "Failed to patch wheel with dynamic version"
+	info, err := patchwheel.ParseWheelFilename(wheel)
+	if err != nil {
+		return "", []diag.Diagnostic{{
+			Severity: diag.Warning,
+			Summary:  msg,
+			Detail:   fmt.Sprintf("When parsing filename \"%s\" encountered an error: %s", wheel, err),
+		}}
+	}
+
+	dir := filepath.Join(cacheDir, "patched_wheels", artifactName+"_"+info.Distribution)
+	createAndCleanupDirectory(ctx, dir)
+	patchedWheel, isBuilt, err := patchwheel.PatchWheel(wheel, dir)
+	if err != nil {
+		return "", []diag.Diagnostic{{
+			Severity: diag.Warning,
+			Summary:  msg,
+			Detail:   fmt.Sprintf("When patching %s encountered an error: %s", wheel, err),
+		}}
+	}
+
+	if isBuilt {
+		log.Infof(ctx, "Patched wheel (built) %s -> %s", wheel, patchedWheel)
+	} else {
+		log.Infof(ctx, "Patched wheel (cache) %s -> %s", wheel, patchedWheel)
+	}
+
+	return patchedWheel, nil
+}
+
+func createAndCleanupDirectory(ctx context.Context, dir string) {
+	err := os.MkdirAll(dir, 0o700)
+	if err != nil {
+		log.Infof(ctx, "Failed to create %s: %s", dir, err)
+		return
+	}
+
+	files, err := os.ReadDir(dir)
+	if err != nil {
+		log.Infof(ctx, "Failed to clean up %s: %s", dir, err)
+		// ReadDir can return partial results, so continue here
+	}
+	for _, entry := range files {
+		path := filepath.Join(dir, entry.Name())
+		err := os.Remove(path)
```

Comment:
This fails for directories. You could call `RemoveAll` to make it work recursively.